src/openforms/submissions/api/validation.py

```diff
@@ -5,65 +5,119 @@
 to the frontend.
 """
 
-from django.core.validators import MaxLengthValidator
+from typing import TYPE_CHECKING, TypedDict
+
 from django.utils.translation import gettext_lazy as _
 
 from rest_framework import serializers
+from rest_framework.exceptions import ErrorDetail
+from rest_framework.request import Request
+from rest_framework.settings import api_settings
 
-from openforms.api.utils import mark_experimental
 from openforms.formio.service import build_serializer, get_dynamic_configuration
+from openforms.forms.models import FormDefinition, FormStep
 
-from ..form_logic import check_submission_logic
 from ..models import Submission, SubmissionStep
-from .fields import (
-    PrivacyPolicyAcceptedField,
-    SubmissionAllowedField,
-    TruthDeclarationAcceptedField,
-)
-
-
-@mark_experimental
-class CompletionValidationSerializer(serializers.Serializer):
-    incomplete_steps = serializers.ListField(
-        child=serializers.CharField(),
-        validators=[
-            MaxLengthValidator(
-                limit_value=0,
-                message=_("Not all applicable steps have been completed: %(value)s"),
-            )
-        ],
+from .fields import PrivacyPolicyAcceptedField, TruthDeclarationAcceptedField
+
+NON_FIELD_ERRORS_KEY = api_settings.NON_FIELD_ERRORS_KEY
+
+
+def is_step_unexpectedly_incomplete(submission_step: SubmissionStep) -> bool:
+    if not submission_step.completed and submission_step.is_applicable:
+        return True
+    return False
+
+
+class SubmissionCompletionSerializerContext(TypedDict):
+    request: Request
+    submission: Submission
+
+
+type StepValidationErrors = dict[str, ErrorDetail] | dict[str, StepValidationErrors]
+
+
+class SubmissionCompletionSerializer(serializers.Serializer):
+    """
+    Validate the submission completion.
+
+    The input validation requires the statement checkboxes to be provided before
+    completion can be considered. If that is fine, we run validation on the state of
+    the submission, which is a bit "weird" in the sense that we're double-checking
+    earlier user-input and not directly validating ``request.data`` items.
+
+    If all is well, we return the status URL to check the background processing,
+    otherwise the validation errors are raised for the frontend to handle.
+    """
+
+    _context: SubmissionCompletionSerializerContext  # pyright: ignore[reportIncompatibleVariableOverride]
+    if TYPE_CHECKING:
+
+        @property
+        def context(  # pyright: ignore[reportIncompatibleMethodOverride]
+            self,
+        ) -> SubmissionCompletionSerializerContext: ...
+
+    # statement checkboxes on the overview page
+    privacy_policy_accepted = PrivacyPolicyAcceptedField(required=False)
+    statement_of_truth_accepted = TruthDeclarationAcceptedField(required=False)
```

Comment:
Why were these changed to required=False?A bit confused with these checkboxes.

Reply:
before, they were required but it doesn't cause any problems because we were manually setting that data to a default (not very different from what we do know in the view), but setting them to required reflects better in the API docs - it doesn't make sense to mark the field as required if the checkboxes are never displayed in the first place.